io/io/inc/ROOT/TBufferMerger.hxx

```diff
@@ -0,0 +1,132 @@
+
+   /** Constructor
+    * @param name Output file name
+    * @param option Output file creation options
+    * @param ftitle Output file title
+    * @param compression Output file compression level
+    */
+   TBufferMerger(const char *name, Option_t *option = "RECREATE", const char *ftitle = "", Int_t compress = 1);
+
+   /** Destructor */
+   ~TBufferMerger();
+
+   /** Returns a TBufferMergerFile to which data can be written.
+    *  At the end, all TBufferMergerFiles get merged into the output file.
+    */
+   std::shared_ptr<TBufferMergerFile> GetFile();
+
+   friend class TBufferMergerFile;
+
+private:
+   void Push(TBufferFile *buffer);
+   void Listen();
+
+   std::mutex fFilesMutex;                                       //< Mutex used to lock fAttachedFiles
+   std::mutex fQueueMutex;                                       //< Mutex used to lock fQueue
+   std::mutex fWriteMutex;                                       //< Mutex used for the condition variable
+   std::condition_variable fCV;                                  //< Condition variable used to wait for data
+   std::queue<TBufferFile *> fQueue;                             //< Queue to which data is pushed and merged
+   std::unique_ptr<TFile> fFile;                                 //< Output file, owned by the user
+   std::unique_ptr<std::thread> fMergingThread;                  //< Worker thread that writes to disk
+   std::vector<std::weak_ptr<TBufferMergerFile>> fAttachedFiles; //< Attached files
+};
+
+/**
+ * \class TBufferMerger TBufferMerger.hxx
+ * \ingroup IO
+ *
+ * A TBufferMergerFile is similar to a TMemFile, but when data
+ * is written to it, it is appended to the TBufferMerger queue.
+ * The TBufferMerger merges all data into the output file on disk.
+ */
+
+class TBufferMergerFile : public TMemFile {
+private:
+   TBufferMerger &fMerger;              //< TBufferMerger this file is attached to
+   std::unique_ptr<TArrayC> fClassSent; //< StreamerInfo this file has already written
+
+   /** Constructor. Can only be called by TBufferMerger.
+    * @param m Merger this file is attached to. */
+   TBufferMergerFile(TBufferMerger &m);
+
+   /** TBufferFile has no default constructor. */
+   TBufferMergerFile() = delete;
+
+   /** TBufferFile has no copy constructor. */
+   TBufferMergerFile(const TBufferMergerFile &) = delete;
+
+   friend class TBufferMerger;
+
+public:
+   /** Destructor */
+   ~TBufferMergerFile();
+
+   using TMemFile::Write;
+
+   /** Write data into a TBufferFile and append it to TBufferMerger.
+    * @param name Name
+    * @param opt  Options
+    * @param bufsize Buffer size
+    * This function must be called before the TBufferMergerFile gets destroyed,
+    * or no data is appended to the TBufferMerger.
+    */
+   virtual Int_t Write(const char *name = nullptr, Int_t opt = 0, Int_t bufsize = 0);
+
+   /** Write StreamerInfo for objects that have not already been written. */
+   void WriteStreamerInfo();
+};
```

Comment:
Since the class inherits from (indirectly) TObject, it must have a ClassDef which could be here
```
   ClassDefOverride(TBufferMergerFile,0);
```
The 0 indicate that the class's data member default to transient.